packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,562 @@
+ */
+type AllowedEvents =
+  | AccountsControllerAccountAddedEvent
+  | AccountsControllerAccountRemovedEvent
+  | AccountsControllerAccountAssetListUpdatedEvent;
+
+/**
+ * Messenger type for the MultichainAssetsController.
+ */
+export type MultichainAssetsControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  MultichainAssetsControllerActions | AllowedActions,
+  MultichainAssetsControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * {@link MultichainAssetsController}'s metadata.
+ *
+ * This allows us to choose if fields of the state should be persisted or not
+ * using the `persist` flag; and if they can be sent to Sentry or not, using
+ * the `anonymous` flag.
+ */
+const assetsControllerMetadata = {
+  assetsMetadata: {
+    persist: true,
+    anonymous: false,
+  },
+  accountsAssets: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+// TODO: make this controller extends StaticIntervalPollingController and update all assetsMetadata once a day.
+
+export class MultichainAssetsController extends BaseController<
+  typeof controllerName,
+  MultichainAssetsControllerState,
+  MultichainAssetsControllerMessenger
+> {
+  // Mapping of CAIP-2 Chain ID to Asset Snaps.
+  #snaps: Record<CaipChainId, Snap[]>;
+
+  readonly #controllerOperationMutex = new Mutex();
+
+  constructor({
+    messenger,
+    state = {},
+  }: {
+    messenger: MultichainAssetsControllerMessenger;
+    state?: Partial<MultichainAssetsControllerState>;
+  }) {
+    super({
+      messenger,
+      name: controllerName,
+      metadata: assetsControllerMetadata,
+      state: {
+        ...getDefaultMultichainAssetsControllerState(),
+        ...state,
+      },
+    });
+
+    this.#snaps = {};
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => await this.#handleOnAccountAddedEvent(account),
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountRemoved',
+      async (account) => await this.#handleOnAccountRemovedEvent(account),
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAssetListUpdated',
+      async (event) => await this.#handleAccountAssetListUpdatedEvent(event),
+    );
+  }
+
+  async #handleAccountAssetListUpdatedEvent(
+    event: AccountAssetListUpdatedEventPayload,
+  ) {
+    return this.#withControllerLock(async () =>
+      this.#handleAccountAssetListUpdated(event),
+    );
+  }
+
+  async #handleOnAccountAddedEvent(account: InternalAccount) {
+    return this.#withControllerLock(async () =>
+      this.#handleOnAccountAdded(account),
+    );
+  }
+
+  /**
+   * Function to update the assets list for an account
+   *
+   * @param event - The list of assets to update
+   */
+  async #handleAccountAssetListUpdated(
+    event: AccountAssetListUpdatedEventPayload,
+  ) {
+    this.#assertControllerMutexIsLocked();
```

Comment:
I'm wondering if it makes sense to throw if the mutex is locked. Why not wait until the mutex is available again instead?

Reply:
Im wrapping the two main fcts `handleOnAccountAddedEvent` and `handleAccountAssetListUpdatedEvent`
using the `withControllerLock` so this `throw` should not be an issue.
This is added as a safe check;  in case `handleAccountAssetListUpdated` is being called in future directly by another fct without using the `withControllerLock`